File: smite/src/main/java/hirez/smite/SmiteGame.java
package hirez.smite;

import hirez.api.*;
import hirez.api.object.*;
import hirez.api.object.interfaces.Queue;
import hirez.api.object.interfaces.ReturnedMessage;
import hirez.smite.object.*;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class SmiteGame extends Endpoint {

    private SmiteGame(Configuration configuration) {
        super(configuration);
    }

    public static SmiteGame create(Consumer<ConfigurationBuilder> configuration) {
        return new SmiteGame(new ConfigurationBuilder().applyFrom((cfg) -> {
            configuration.accept(cfg);
            if (cfg.getBaseEndpoint() == null) {
                cfg.setBaseEndpoint(SmitePlatform.PC);
            }
        }).build());
    }

    public Flowable<GodLeaderboard> getGodLeaderboard(long godId, Queue queue) {
        return testAndCall(GodLeaderboard[].class, "getgodleaderboard", Long.toString(godId), queue.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<ItemRecommendation> getGodRecommendedItems(long godId, Language language) {
        return testAndCall(ItemRecommendation[].class, "getgodrecommendeditems", Long.toString(godId), language.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<ItemRecommendation> getGodRecommendedItems(long godId) {
        return getGodRecommendedItems(godId, getConfiguration().getLanguage());
    }

    public Flowable<God> getGods(Language language) {
        return testAndCall(God[].class, "getgods", language.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<God> getGods() {
        return getGods(getConfiguration().getLanguage());
    }

    public Flowable<GodSkin> getGodSkins(long godId, Language language) {
        return testAndCall(GodSkin[].class, "getgodskins", Long.toString(godId), language.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<GodSkin> getGodSkins(long godId) {
        return getGodSkins(godId, getConfiguration().getLanguage());
    }

    public Flowable<Item> getItems(Language language) {
        return testAndCall(Item[].class, "getitems", language.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<Item> getItems() {
        return getItems(getConfiguration().getLanguage());
    }

    public Single<DemoDetail> getDemoDetails(long matchId) {
        return testAndCall(DemoDetail[].class, "getdemodetails", Long.toString(matchId))
                .map(it -> it[0]);
    }

    public Flowable<LeagueLeaderboard> getLeagueLeaderboard(Queue queue, Division division, int season) {
        return testAndCall(LeagueLeaderboard[].class, "getleagueleaderboard", queue.getId().toString(), Integer.toString(division.ordinal()), Integer.toString(season))
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<LeagueSeason[]> getLeagueSeasons(Queue queue) {
        return testAndCall(LeagueSeason[].class, "getleagueseasons", queue.getId().toString());
    }

    public Flowable<MatchDetail> getMatchDetails(long matchId) {
        return testAndCall(MatchDetail[].class, "getmatchdetails", Long.toString(matchId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<MatchDetail> getMatchDetailsBatch(long... matchId) {
        if (matchId.length < 5) {
            return Flowable.error(new HiRezException("To get Batched matches you must defined minimum 5 matchId's"));
        }
        return testAndCall(MatchDetail[].class, "getmatchdetailsbatch", Arrays.stream(matchId).mapToObj(Long::toString).collect(Collectors.joining(",")))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<MatchHistory> getMatchHistory(long userId) {
        return testAndCall(MatchHistory[].class, "getmatchhistory", Long.toString(userId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<MatchId> getMatchIdsByQueue(Queue queue, Date timestamp) {
        DateFormat df = new SimpleDateFormat("yyyyMMdd/HH");
        DateFormat mdf = new SimpleDateFormat("mm");

        int minutes = Integer.parseInt(mdf.format(timestamp));

        minutes = minutes - minutes % 10 + ((minutes % 10) > 4 ? 10 : 0);

        return testAndCall(MatchId[].class, "getmatchidsbyqueue", queue.getId().toString(), df.format(timestamp) + minutes)
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<LiveMatch> getMatchPlayerDetails(long matchId) {
        return testAndCall(LiveMatch[].class, "getmatchplayerdetails", Long.toString(matchId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<QueueStat> getQueueStats(long userId, Queue queue) {
        return testAndCall(QueueStat[].class, "getqueuestats", Long.toString(userId), queue.getId().toString())
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<TopMatch> getTopMatches() {
        return testAndCall(TopMatch[].class, "gettopmatches")
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<Friend> getFriends(long userId) {
        return testAndCall(Friend[].class, "getfriends", Long.toString(userId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<GodRank> getGodRanks(long userId) {
        return testAndCall(GodRank[].class, "getgodranks", Long.toString(userId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<Player> getPlayer(long userId) {
        return testAndCall(Player[].class, "getplayer", Long.toString(userId))
                .map(it -> it[0]);
    }

    public Single<Player> getPlayer(long userId, Portal portal) {
        return testAndCall(Player[].class, "getplayer", Long.toString(userId), portal.getId().toString())
                .map(it -> it[0]);
    }

    public Single<Achievements> getPlayerAchievements(long userId) {
        return testAndCall(Achievements.class, "getplayerachievements", Long.toString(userId));
    }

    public Flowable<PlayerIdPortal> getPlayerIdByName(String query) {
        return testAndCall(PlayerIdPortal[].class, "getplayeridbyname", query)
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<PlayerIdPortal> getPlayerIdByPortalUserId(Portal portal, long id) {
        return testAndCall(PlayerIdPortal[].class, "getplayeridbyportaluserid", portal.getId().toString(), Long.toString(id))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<PlayerIdPortal> getPlayerIdsByGamerTag(Portal portal, String query) {
        return testAndCall(PlayerIdPortal[].class, "getplayeridsbygamertag", portal.getId().toString(), query)
                .flattenAsFlowable(Arrays::asList);
    }

    public Single<PlayerStatus> getPlayerStatus(long userId) {
        return testAndCall(PlayerStatus[].class, "getplayerstatus", Long.toString(userId))
                .map(it -> it[0]);
    }

    public Single<Team> getTeamDetails(long teamId) {
        return testAndCall(Team[].class, "getteamdetails", Long.toString(teamId))
                .map(it -> it[0]);
    }

    public Flowable<TeamMember> getTeamPlayers(long teamId) {
        return testAndCall(TeamMember[].class, "getteamplayers", Long.toString(teamId))
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<TeamQuery> searchTeams(String query) {
        return testAndCall(TeamQuery[].class, "searchteams", query)
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<ProLeagueDetail> getEsportsProLeagueDetails() {
        return testAndCall(ProLeagueDetail[].class, "getesportsproleaguedetails")
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<MatchOfTheDay> getMOTD() {
        return testAndCall(MatchOfTheDay[].class, "getmotd")
                .flattenAsFlowable(Arrays::asList);
    }

    public Flowable<PlayerQuery> searchPlayer(String query) {
        return testAndCall(PlayerQuery[].class, "searchplayers", query)
                .flattenAsFlowable(Arrays::asList);
    }

    static boolean isRanked(int id) {
        return Arrays.asList(440, 450, 451, 502, 503, 504).contains(id);
    }
}

File: smite/src/main/kotlin/hirez/smite/SmiteGame.kt
package hirez.smite

import com.google.gson.Gson
import hirez.Config
import hirez.ConfigBuilder
import hirez.Language
import hirez.api.AbstractAPI
import hirez.api.Http
import hirez.api.Queue
import hirez.api.exceptions.PlayerNotFoundException
import hirez.enums.Portal
import hirez.flattenArray
import hirez.json.CreateSession
import hirez.json.DataUsage
import hirez.json.Item
import hirez.json.Patch
import hirez.json.Ping
import hirez.json.PlayerData
import hirez.json.Server
import hirez.json.TestSession
import hirez.smite.json.God
import hirez.smite.json.GodLeaderboard
import hirez.smite.json.GodSkin
import hirez.smite.json.Player
import hirez.statuspage.StatusPage
import io.reactivex.Flowable
import io.reactivex.Single

/**
 * @author Damian Staszewski [damian@stachuofficial.tv]
 * @version %I%, %G%
 * @since 1.0
 */
object SmiteGame : AbstractAPI() {
	@JvmStatic
	fun initConfig(config: Config) {
		this.config = config
		this.http = Http(Gson(), config.userAgent)
	}
	
	@JvmStatic
	fun initConfig(configBuilder: ConfigBuilder.() -> Unit) =
				initConfig(ConfigBuilder().apply(configBuilder).build())
	
	
	@JvmStatic
	fun createsession() = get<CreateSession>("createsession").doOnSuccess(config.sessionStorage::set)
	val dataused
		@JvmStatic get() = get<DataUsage>("getdataused")
	val hirezserverstatus
		@JvmStatic get() = get<Array<Server>>("gethirezserverstatus")
	val patchinfo
		@JvmStatic get() = get<Patch>("getpatchinfo")
	
	@JvmStatic
	fun ping() = get<String>("ping").map(::Ping)
	
	@JvmStatic
	fun testsession() = get<String>("testsession")
				.map(::TestSession)
				.doOnSuccess { if (!it.isSuccessful) config.sessionStorage.remove() }
	
	fun getGodLeaderboard(godId: Long, queue: Queue) =
				Single.just(queue).flatMap { if (queue.isRanked) get<Array<GodLeaderboard>>("getgodleaderboard") else Single.error(IllegalArgumentException("Only ranked games are be allowed")) }
							.flattenArray()
	fun getGodRecommendedItems() =
				get<Array<Item>>("getgodrecommendeditems")
							.flattenArray()
	fun getGods() =
				get<Array<God>>("getgods")
							.flattenArray()
	fun getGodSkins() =
				get<Array<GodSkin>>("getgodskins")
							.flattenArray()
	fun getItems() =
				get<Array<Item>>("getitems")
							.flattenArray()
	fun getDemoDetails() =
				get<>("getdemodetails")
	fun getLeagueLeaderboard() =
				get<>("getleagueleaderboard")
	fun getLeagueSeasons() =
				get<>("getleagueseasons")
	fun getMatchDetails() =
				get<>("getmatchdetails")
	fun getMatchDetailsBatch() =
				get<>("getmatchdetailsbatch")
	fun getMatchHistory() =
				get<>("getmatchhistory")
	fun getMatchIdsByQueue() =
				get<>("getmatchidsbyqueue")
	fun getMatchPlayerDetails() =
				get<>("getmatchplayerdetails")
	fun getQueueStats() =
				get<>("getqueuestats")
	fun getTopMatches() =
				get<>("gettopmatches")
	fun getFriends() =
				get<>("getfriends")
	fun getGodRanks() =
				get<>("getgodranks")
	fun getPlayer() =
				get<>("getplayer")
	fun getPlayerAchievements() =
				get<>("getplayerachievements")
	fun getPlayerIdByName() =
				get<>("getplayeridbyname")
	fun getPlayerIdByPortalUserId() =
				get<>("getplayeridbyportaluserid")
	fun getPlayerIdsByGamertag() =
				get<>("getplayeridsbygamertag")
	fun getPlayerStatus() =
				get<>("getplayerstatus")
	fun getTeamDetails() =
				get<>("getteamdetails")
	fun getTeamPlayers() =
				get<>("getteamplayers")
	fun searchTeams() =
				get<>("searchteams")
	fun getEsportsProLeagueDetails() =
				get<>("getesportsproleaguedetails")
	fun getMOTD() =
				get<>("getmotd")

//	@JvmStatic
//	fun ping() = get<String>("ping")
//				.map(::Ping)
//
//	@JvmStatic
//	fun createSession() =
//				get<CreateSession>("createsession")
//							.doOnSuccess(config.sessionStorage::set)
//
//	@JvmStatic
//	fun testSession() = get<String>("testsession")
//				.map(::TestSession)
//				.doOnSuccess { if (!it.isSuccessful) config.sessionStorage.remove() }
//
//	@JvmStatic
//	val dataUsed
//		get() = get<Array<DataUsage>>("getdataused")
//					.map { it[0] }
//
//	@JvmStatic
//	val patchInfo
//		get() = get<Patch>("getpatchinfo")
//
//	@JvmStatic
//	val statusPage
//		get() = StatusPage(http)
//
//	@JvmStatic
//	val hiRezServerStatus
//		get() = get<Array<Server>>("gethirezserverstatus")
//					.flattenAsFlowable { it.asIterable() }
//	@JvmStatic
//	val gods
//		get() = getGods(config.defaultLanguage)
//
//	@JvmStatic
//	fun getGods(language: Language) =
//				get<Array<God>>("getgods", language.id.toString())
//							.flattenAsFlowable { it.asIterable() }
//
//	@JvmStatic
//	fun getGodLeaderboard(godId: Long, queue: Queue) = get<Array<GodLeaderboard>>("getgodleaderboard", godId.toString(), queue.id.toString())
//				.flattenAsFlowable { it.asIterable() }
//
//	@JvmStatic
//	fun getGodSkins(godId: Long, language: Language) =
//				get<Array<GodSkin>>("", godId.toString(), language.id.toString())
//							.flattenAsFlowable { it.asIterable() }
//
//	@JvmStatic
//	fun getGodSkins(godId: Long) = SmiteGame.getGodSkins(godId, config.defaultLanguage)
//
//	@JvmStatic
//	fun getGodRecommendedItems(godId: Long, language: Language) =
//				get<Array<Item>>("getgodrecommendeditems", godId.toString(), language.id.toString())
//							.flattenAsFlowable { it.asIterable() }
//
//	@JvmStatic
//	fun getGodRecommendedItems(godId: Long) = getGodRecommendedItems(godId, config.defaultLanguage)
//
//	@JvmStatic
//	fun getItems(language: Language) =
//				get<Array<Item>>("getitems", language.id.toString())
//							.flattenAsFlowable { it.asIterable() }
//
//	@JvmStatic
//	val items
//		get() = getItems(config.defaultLanguage)
//
//	@JvmStatic
//	fun getPlayer(name: String) =
//				getPlayerIdByName(name).flatMap { get<Array<Player>>("getplayer", it[0].id.toString()) }
//							.map { if (it.isNotEmpty()) it[0] else throw PlayerNotFoundException("This profile is hidden: $name") }
//
//	@JvmStatic
//	fun getPlayer(name: String, portal: Portal) =
//				getPlayerIdByName(name).flatMap {
//					val p = it.firstOrNull { it.portal == portal }
//					if (p != null) get<Array<Player>>("getplayer", p.id.toString(), p.portal.id.toString())
//					else Single.error(PlayerNotFoundException("Player: $name is not assignable to this portal: ${portal.name}"))
//				}.map { if (it.isNotEmpty()) it[0] else throw PlayerNotFoundException("This profile is hidden: $name") }
//
//	@JvmStatic
//	fun getPlayerIdByName(name: String) =
//				get<Array<PlayerData>>("getplayeridbyname", name)
//							.map { if (it.isNotEmpty()) it else throw PlayerNotFoundException("Player does not exist: $name") }
//
//	@JvmStatic
//	fun getPlayerIdByPortalUserId(userId: Long, portal: Portal) =
//				get<Array<PlayerData>>("getplayeridbyportaluserid", portal.id.toString(), userId.toString())
//							.map { if (it.isNotEmpty()) it else throw PlayerNotFoundException("Player does not exist: ID[$userId]") }
//
//	@JvmStatic
//	fun getPlayerIdByGamertag(gamertag: String, portal: Portal) =
//				get<Array<PlayerData>>("getplayeridbyportaluserid", portal.id.toString(), gamertag)
//							.map { if (it.isNotEmpty()) it else throw PlayerNotFoundException("Player does not exist: $gamertag") }
	
}